crates/storage/src/interface.rs

```diff
@@ -44,8 +64,31 @@ impl Interface {
     /// If an error occurs when interacting with the storage system, an error
     /// will be returned.
     ///
-    pub fn find_by_id(&self, _id: Id) -> Result<Option<Element>, StorageError> {
-        unimplemented!()
+    #[allow(clippy::significant_drop_tightening)]
+    pub fn find_by_id(&self, id: Id) -> Result<Option<Element>, StorageError> {
+        // TODO: It seems fairly bizarre/unexpected that the put() method is sync
+        // TODO: and not async. The reasons and intentions need checking here, in
+        // TODO: case this find() method should be async and wrap the blocking call
+        // TODO: with spawn_blocking(). However, this is not straightforward at
+        // TODO: present because Slice uses Rc internally for some reason.
```

Comment:
rocksdb doesn't expose an async interface, so to keep it simple, we didn't either

that, plus, async traits were only recently stabilized, and still lack proper support for dynamic dispatch, easy mitigation would've been to use the `async_trait` crate, but that comes with it's own issues

so, we kept it simple

as for the `Rc` bit, a mitigation to that would've made it in https://github.com/calimero-network/core/pull/519 ([src](https://github.com/calimero-network/core/pull/519)) but there was no time for the ripple effect of code change required, it's now planned at a later date.

Reply:
Okay, that makes sense. Just a couple of thoughts, though:

  - Although RocksDB is sync, we may want to consider making our calls to it async, to avoid blocking generally, as this will become a scalability concern.
  - Async traits stabilised fairly recently (end of last year from memory...?) but we've always had the functionality; we just had to roll the futures by hand or use the `async_trait` crate. So the lack of stabilisation in Rust core didn't really prevent anything here.
  - We could trivially change the `Rc` to an `Arc` couldn't we...?

I am wondering, do you think we should change the `Rc` to an `Arc` to allow `async` here, which then avoids thread blocking?

Reply:
> So the lack of stabilisation in Rust core didn't really prevent anything here.

https://github.com/calimero-network/core/blob/1db5a74d5afd88d294c10a5b758cdae383db0cbd/crates/store/src/lib.rs#L22-L24

`Store`'s is implemented with dynamic dispatch, neither of these are object safe since the returned future from different implementations might be different types:

```rust
trait Database {
    fn get(&self) -> impl Future<Output = ()>;
}
```

```rust
trait Database {
    async fn get(&self);
}
```

we'd have had to forcefully unify the return type to a known `Box<dyn Future<Output = ()>>` to abstract away that variance or introduce `async_trait`, and at the point of implementation, I decided against it, we'll do this in a future effort where we can incorporate https://github.com/calimero-network/core/compare/455fe09ca9be09df17046584a3ef6cd28564e01a~...f74f4f48e39a8ad2556881c4a0f599fba346f766

> I am wondering, do you think we should change the Rc to an Arc to allow async here, which then avoids thread blocking?

for now, you have to think about where this will eventually be used, as Wasmer, our VM, doesn't have an async interface, and as such, handling host calls from the guest cannot use any async interfaces without blocking anyway

unless there's some immediate, critical need to depend on something that only exports an async interface, I'd say we keep it simple and adopt the sync model instead

once we sort out async execution for the runtime, and the store, we'll do this too

I'd earmarked to resolve those issues during the implementation of the task manager ([ref](https://github.com/calimero-network/core/issues/188)), but there's a lot more questions to be answered surrounding that

Reply:
That's absolutely fine, and I appreciate the insightful response 🙂